library/psa_crypto_storage_its.c

```diff
@@ -27,6 +27,7 @@
 
 #if defined(MBEDTLS_PSA_CRYPTO_STORAGE_ITS_C)
 
+#include "psa/error.h"
```

Comment:
Not sure why this would fail the build. Type definitions in both files are the same, meaning that the compiler should not fail this duplication. I build it this way in mbed-os without issues.
Is this an empirical or a theoretical problem? I ask because I only added `psa/error.h` to my current mbed-os PR, so if this is empirical, than where do you take `psa/error.h` from?

Reply:
Theoretical in this instance, but empirical from a previous version. We had to add guards around the definition of `psa_status_t` so that it would only be defined in one place. You can have multiple `#define` for the same macro as long as they expand to the same token sequence, but some compilers warn about it. You can't have multiple `typedef` statements defining the same type name, even if the definitions are identical.

In your build, what exactly does `psa/error.h` contain? Who's defining `psa_status_t` and how do you avoid defining it twice?

Reply:
`psa/error.h` is part of the PSA include directory. I'm adding it now to [the parallel PR in mbed-os](https://github.com/ARMmbed/mbed-os/pull/9708). Please note that `psa_status_t` is not a define but a typedef. As opposed to defines, compilers will not warn about duplicate typedefs as long as the referenced type is identical. In addition, as it's a typedef and not a define, one can't ask whether it's defined using `#ifdef`.

Reply:
Duplicate definitions of a typedef _are_ forbidden in C89 and C99. From N1256 (C99 TC3), §6.7 “Declarations”:

> 3\. If an identifier has no linkage, there shall be no more than one declaration of the identifier
(in a declarator or type specifier) with the same scope and in the same name space, except
for tags as specified in 6.7.2.3.

Clang does warn about it if you ask it to use C89 or C99.
```
$ cat a.c
typedef int foo;
typedef int foo;
$ clang -std=c99 -Wall -Wextra -Werror -O -c a.c
a.c:2:13: error: redefinition of typedef 'foo' is a C11 feature
      [-Werror,-Wtypedef-redefinition]
typedef int foo;
            ^
a.c:1:13: note: previous definition is here
typedef int foo;
            ^
1 error generated.
```
Neither Clang nor GCC warn about it in non-picky mode, but I remember we've hard this error in at least one of the CC+CFLAGS combination we test in our CI, and we do aim to stick to standard C in Mbed TLS and Mbed Crypto.

Reply:
OK, understood (didn't go that far).
How do you suggest to resolve this? It's a bit of a pickle, as there's no `#ifdef` for typedefs. I could ask whether `PSA_SUCCESS` is defined, assuming that `PSA_SUCCESS` and `psa_status_t` are both defined in `psa/error.h`, but that just seems pretty awkward to me.

Reply:
That's the hack we had for a long time, and getting rid of that hack was my motivation for requesting a unified `psa/error.h` (https://github.com/ARMmbed/PSA-IPC-doc/issues/133) as the only place where `psa_status_t` is defined. Crypto is planning to switch from defining `psa_status_t` in `crypto_types.h` to including `error.h` as soon as version 1 of `error.h` has stabilized.

Reply:
OK, will add this hack.